### PR TITLE
Add some explicit casts to IndexSet.

### PR DESCRIPTION
**Pull request by drwells**
I tried adding an array PetscInts to an index set and got a sign conversion warning.

### CODE REVIEW

**Comment by drwells** [comment]
```
/Users/runner/work/dealii/dealii/bundled/tbb-2018_U2/include/tbb/tbb_stddef.h:318:10: error: definition of implicit copy constructor for 'no_assign' is deprecated because it has a user-declared copy assignment operator [-Werror,-Wdeprecated-copy]
    void operator=( const no_assign& );
```
👀

File: include/deal.II/base/index_set.h [hunk 1]
```diff
@@ -1723,7 +1723,7 @@ IndexSet::add_indices(const ForwardIterator &begin, const ForwardIterator &end)
       size_type       end_index   = begin_index + 1;
       ForwardIterator q           = p;
       ++q;
-      while ((q != end) && (*q == end_index))
+      while ((q != end) && (size_type(*q) == end_index))
```
Comment (tjhei) [inline include/deal.II/base/index_set.h:1726]: why not do the following?
```suggestion
      while ((q != end) && (static_cast<size_type>(*q) == end_index))
```

Comment (bangerth) [reply include/deal.II/base/index_set.h:1726]: Yes, I think the explicit cast is what you want.

**Review by bangerth — APPROVED** [review]

**Pull request merged**